Packaging dracut

**Project description**

dracut enhances the initramfs/initrd. Primarily I was looking for it userspace tools like `lsinitrd`.

Official github repo: https://github.com/dracutdevs/dracut (it might be a kernel mirror, or the kernel might be a mirror of the github, not sure).

Anyway my progress in trying to package this was:

1. Create shell.nix

```
{ pkgs ? import <nixpkgs> {} }:

with pkgs;
stdenv.mkDerivation rec {
  name = "dracut";
  src = null;
  nativeBuildInputs = [ pkg-config ];
  shellHook = ''
    echo 'Entering ${name}'
  '';
}
```

2. Then run `patchShebangs ./configure` (it has /bin/bash)

3. Then run ./configure

The result is:

```
Package libkmod was not found in the pkg-config search path.
Perhaps you should add the directory containing `libkmod.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libkmod' found
dracut needs pkg-config and libkmod >= 23.
```

I cannot find `libkmod` and I don't know where to even get that.

With this default.nix:

```nix
{ pkgs ? import <nixpkgs> {} }:

with pkgs;
stdenv.mkDerivation rec {
  name = "dracut";
  src = ./.;
  nativeBuildInputs = [
    pkg-config
    kmod
    asciidoc
    libxslt
  ];
  preConfigure = ''
    patchShebangs ./configure
  '';
}
```

I'm getting now:

```
[nix-shell:~/Projects/dracut]$ nix-build
these derivations will be built:
  /nix/store/ng7r9z0aamqvi2b9qm55g1mvdj1v3xf7-dracut.drv
building '/nix/store/ng7r9z0aamqvi2b9qm55g1mvdj1v3xf7-dracut.drv'...
unpacking sources
unpacking source archive /nix/store/j5y47p9nwnpr9lwqbl2w8hwqbzhyj4s6-dracut
source root is dracut
patching sources
configuring
patching script interpreter paths in ./configure
./configure: interpreter directive changed from "/bin/bash" to "/nix/store/506nnycf7nk22x7n07mjjjl2g8nifpda-bash-4.4-p23/bin/bash"
configure flags: --prefix=/nix/store/gkp5sq4qlwwf48q3c5n186nn73af51zb-dracut
building
build flags: SHELL=/nix/store/506nnycf7nk22x7n07mjjjl2g8nifpda-bash-4.4-p23/bin/bash
gcc -O2 -g -Wall -std=gnu99 -D_FILE_OFFSET_BITS=64 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -I/nix/store/7wg0l9szgl79jn1dqghr2rzxwsns97pq-kmod-26/include   -c -o install/dracut-install.o install/dracut-install.c
gcc -O2 -g -Wall -std=gnu99 -D_FILE_OFFSET_BITS=64 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -I/nix/store/7wg0l9szgl79jn1dqghr2rzxwsns97pq-kmod-26/include   -c -o install/hashmap.o install/hashmap.c
gcc -O2 -g -Wall -std=gnu99 -D_FILE_OFFSET_BITS=64 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -I/nix/store/7wg0l9szgl79jn1dqghr2rzxwsns97pq-kmod-26/include   -c -o install/log.o install/log.c
gcc -O2 -g -Wall -std=gnu99 -D_FILE_OFFSET_BITS=64 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -I/nix/store/7wg0l9szgl79jn1dqghr2rzxwsns97pq-kmod-26/include   -c -o install/strv.o install/strv.c
gcc -O2 -g -Wall -std=gnu99 -D_FILE_OFFSET_BITS=64 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -I/nix/store/7wg0l9szgl79jn1dqghr2rzxwsns97pq-kmod-26/include   -c -o install/util.o install/util.c
gcc  -o install/dracut-install install/dracut-install.o install/hashmap.o install/log.o install/strv.o install/util.o  -L/nix/store/7wg0l9szgl79jn1dqghr2rzxwsns97pq-kmod-26/lib -lkmod
ln -fs install/dracut-install dracut-install
gcc -O2 -g -Wall -std=gnu99 -D_FILE_OFFSET_BITS=64 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -I/nix/store/7wg0l9szgl79jn1dqghr2rzxwsns97pq-kmod-26/include   -c -o skipcpio/skipcpio.o skipcpio/skipcpio.c
gcc   skipcpio/skipcpio.o   -o skipcpio/skipcpio
asciidoc -d manpage -b docbook -o "lsinitrd.1.xml" lsinitrd.1.asc
xsltproc -o "lsinitrd.1" -nonet http://docbook.sourceforge.net/release/xsl/current/manpages/docbook.xsl lsinitrd.1.xml
I/O error : Attempt to load network entity http://docbook.sourceforge.net/release/xsl/current/manpages/docbook.xsl
warning: failed to load external entity "http://docbook.sourceforge.net/release/xsl/current/manpages/docbook.xsl"
cannot parse http://docbook.sourceforge.net/release/xsl/current/manpages/docbook.xsl
make: *** [Makefile:91: lsinitrd.1] Error 4
rm lsinitrd.1.xml
builder for '/nix/store/ng7r9z0aamqvi2b9qm55g1mvdj1v3xf7-dracut.drv' failed with exit code 2
error: build of '/nix/store/ng7r9z0aamqvi2b9qm55g1mvdj1v3xf7-dracut.drv' failed
```

Obviously can't load that foreign resource during build. But how to disable?

--- a/pkgs/os-specific/linux/dracut/default.nix
+++ b/pkgs/os-specific/linux/dracut/default.nix
@@ -1,0 +1,128 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, gitUpdater
+, makeBinaryWrapper
+, pkg-config
+, asciidoc
+, libxslt
+, docbook_xsl
+, bash
+, kmod
+, binutils
+, busybox
+, bzip2
+, coreutils
+, cpio
+, findutils
+, glibc
+, gnugrep
+, gnused
+, gnutar
+, gzip
+, kbd
+, lvm2
+, lz4
+, lzop
+, procps
+, rng-tools
+, squashfsTools
+, systemd
+, util-linux
+, xz
+, zstd
+}:
+
+stdenv.mkDerivation rec {
+  pname = "dracut";
+  version = "059";
+
+  src = fetchFromGitHub {
+    owner = "dracutdevs";
+    repo = "dracut";
+    rev = version;
+    hash = "sha256-zSyC2SnSQkmS/mDpBXG2DtVVanRRI9COKQJqYZZCPJM=";
+  };
+
+  strictDeps = true;
+
+  buildInputs = [
+    bash
+    kmod
+  ];
+
+  nativeBuildInputs = [
+    makeBinaryWrapper
+    pkg-config
+    asciidoc
+    libxslt
+    docbook_xsl
+  ];
+
+  postPatch = ''
+    substituteInPlace dracut.sh \
+      --replace 'dracutbasedir="$dracutsysrootdir"/usr/lib/dracut' 'dracutbasedir="$dracutsysrootdir"'"$out/lib/dracut"
+    substituteInPlace lsinitrd.sh \
+      --replace 'dracutbasedir=/usr/lib/dracut' "dracutbasedir=$out/lib/dracut"
+
+    echo 'DRACUT_VERSION=${version}' >dracut-version.sh
+  '';
+
+  preConfigure = ''
+    patchShebangs ./configure
+  '';
+
+  postFixup = ''
+    wrapProgram $out/bin/dracut --prefix PATH : ${lib.makeBinPath [
+      coreutils
+      util-linux
+    ]} --prefix DRACUT_PATH : ${lib.makeBinPath [
+      bash
+      binutils
+      coreutils
+      findutils
+      glibc
+      gnugrep
+      gnused
+      gnutar
+      kbd
+      lvm2
+      procps
+      rng-tools
+      squashfsTools
+      systemd
+      util-linux
+      busybox
+    ]}
+    wrapProgram $out/bin/dracut-catimages --set PATH ${lib.makeBinPath [
+      coreutils
+      cpio
+      findutils
+      gzip
+    ]}
+    wrapProgram $out/bin/lsinitrd --set PATH ${lib.makeBinPath [
+      binutils
+      bzip2
+      coreutils
+      cpio
+      gnused
+      gzip
+      lz4
+      lzop
+      squashfsTools
+      util-linux
+      xz
+      zstd
+    ]}
+  '';
+
+  passthru.updateScript = gitUpdater { };
+
+  meta = with lib; {
+    homepage = "https://dracut.wiki.kernel.org";
+    description = "An event driven initramfs infrastructure";
+    license = licenses.gpl2Plus;
+    maintainers = with maintainers; [ lilyinstarlight ];
+    platforms = platforms.linux;
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -28399,6 +28399,8 @@ with pkgs;
 
   dr14_tmeter = callPackage ../applications/audio/dr14_tmeter { };
 
+  dracut = callPackage ../os-specific/linux/dracut { };
+
   dragonflydb = callPackage ../servers/nosql/dragonflydb { };
 
   dragonfly-reverb = callPackage ../applications/audio/dragonfly-reverb { };

